src/fastfilter.h

```diff
@@ -43,47 +45,74 @@ class CFastFilter
 
     CFastFilter()
     {
+        static_assert((HASHES > 1) && (HASHES <= 16), "HASHES must be between 2 and 16 inclusive");
         FastRandomContext insecure_rand;
         vData.resize(FILTER_BYTES);
-        // by sampling from a different part of the uint256, we make it harder for an attacker to generate collisions
-        grabFrom = (1 + (insecure_rand.rand32() % 6)) * 4; // should be 4 byte aligned for speed
+        GetRandBytes(salt.begin(), sizeof(uint256));
     }
 
     // returns true IF this function made a change (i.e. the value was previously not set).
     bool checkAndSet(const uint256 &hash)
     {
-        const unsigned char *mem = hash.begin();
-        const uint32_t *pos = (const uint32_t *)&(mem[grabFrom]);
-        uint32_t idx = ((*pos) ^ mem[0]) & (FILTER_SIZE - 1);
-
-        bool ret = vData[idx >> 3] & (1 << (idx & 7));
-        vData[idx >> 3] |= (1 << (idx & 7));
-        if (ret)
-            conflicts++;
-        return !ret;
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        bool unset = 0; // If any position is not set, then this will be true
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
+            uint32_t idx2 = val & (FILTER_SIZE - 1);
+
+            unset |= (0 == (vData[idx >> 3] & (1 << (idx & 7))));
+            unset |= (0 == (vData[idx2 >> 3] & (1 << (idx2 & 7))));
+
+            vData[idx >> 3] |= (1 << (idx & 7));
+            vData[idx2 >> 3] |= (1 << (idx2 & 7));
+        }
+        return unset;
     }
 
     void insert(const uint256 &hash)
     {
-        const unsigned char *mem = hash.begin();
-        const uint32_t *pos = (const uint32_t *)&(mem[grabFrom]);
-        uint32_t idx = ((*pos) ^ mem[0]) & (FILTER_SIZE - 1);
-        vData[idx >> 3] |= (1 << (idx & 7));
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
+            uint32_t idx2 = val & (FILTER_SIZE - 1);
+
+            vData[idx >> 3] |= (1 << (idx & 7));
+            vData[idx2 >> 3] |= (1 << (idx2 & 7));
+        }
```

Comment:
I want to rely as little as possible on compiler optimization, since its easy for such to silently disappear based on small tangentially related changes.

Reply:
How about I rework this to use `inline` and you try to find such a small tangentially related change that will screw it up?

Compilers are quite good nowadays :-)

Reply:
They aren't.  I could probably get another 2 to 10x improvement if I used assembly.  This is why libsecp is asm.  Note also that inlining is optional in the compiler.